Feat/spring security 기능 구현 #51

kendrickhama opened the pull request:
## 📌 PR 개요
- spring security

## #️⃣연관된 이슈

> #51 

## 📝작업 내용
- JWT 생성 및 로그인
- access&refresh token
- 로그아웃 기능 구현

kendrickhama committed:
🚧 [WIP] JWT 생성 로직 추가 및 검증 작업 진행 중

kendrickhama committed:
✨ [feat] JWTFilter 기능 추가

kendrickhama committed:
♻️ [refactor] UserRepository findByEmail 수정

kendrickhama committed:
merge 후 충돌 해결

kendrickhama committed:
✨ [feat] Refresh 토큰, Rotate, 로그아웃 기능 추가

kendrickhama committed:
✨ [feat] RoleHierarchy 설정 추가

kendrickhama committed:
Merge branch 'dev' into feat/spring-security-기능-구현-#51

kendrickhama committed:
♻️ [refactor] setStatus 코드수정

ziguin0925 reviewed (commented):
리뷰 확인 부탁드립니다.

ziguin0925 commented on src/main/java/com/investmetic/domain/user/controller/InvestorController.java:
```diff
@@ -0,0 +1,31 @@
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+public class InvestorController {
```
해당 컨트롤러 사용안해서 빼도 괜찮을 것 같아요

ziguin0925 commented on src/main/java/com/investmetic/domain/user/controller/ReIssueController.java:
```diff
@@ -0,0 +1,83 @@
+        String role = jwtUtil.getRole(refresh);
+
+        //새 토큰 발급
+        String newAccess = jwtUtil.createJwt("access", username, role, 30 * 60 * 1000L); //30분
```
@Value ...

ziguin0925 reviewed (commented):
리뷰 다시 올렸습니당

kendrickhama committed:
♻️ [refactor] 리뷰 내용 반영완료

kendrickhama committed:
♻️ [refactor] 리뷰 내용 반영완료

ziguin0925 reviewed (dismissed):
병훈님께도 pr 한번 말씀해주시고 진행해주세요. yaml파일 공유도...

kbyunghoon commented on src/main/java/com/investmetic/global/security/handler/CustomAuthenticationSuccessHandler.java:
```diff
@@ -0,0 +1,78 @@
+
+        // 응답 헤더와 JSON 설정
+        response.setHeader("access", "Bearer " + access);
+        response.addCookie(createCookie("refresh", refresh));
```
리프레시 토큰을 Redis에 저장할 거면 해당 리프레시 토큰을 클라이언트에 넘길 필요 없습니다.

kbyunghoon commented on src/main/java/com/investmetic/global/util/JWTUtil.java:
```diff
@@ -0,0 +1,74 @@
+
+    private final SecretKey secretKey;
+
+    public JWTUtil(@Value("${jwt.secret}") String secret) {
```
Value 따로 빼주세요

kbyunghoon commented on src/main/java/com/investmetic/global/util/RedisUtil.java:
```diff
@@ -29,4 +31,16 @@ public void setDataExpire(String key, String value, long duration) {
     public void deleteData(String key) {
         template.delete(key);
     }
+
+    public void saveRefreshToken(String username, String refreshToken, long durationInSeconds) {
+        template.opsForValue().set("REFRESH_TOKEN:" + username, refreshToken, durationInSeconds, TimeUnit.SECONDS);
+    }
+
+    public String getRefreshToken(String username) {
+        return template.opsForValue().get("REFRESH_TOKEN:" + username);
+    }
+
+    public void deleteRefreshToken(String username) {
+        template.delete("REFRESH_TOKEN:" + username);
+    }
```
여기도 확인해주세요

kbyunghoon reviewed (changes requested):
리뷰 확인해주세요

kendrickhama committed:
♻️ [refactor] ReissueController. 서비스 따로 만들어서 커스텀 예외처리 완료

kendrickhama committed:
♻️ [refactor] 사용 안 하는 메서드 제거

kendrickhama committed:
♻️ [refactor] ReissueController에서 DTO로 반환하게끔 DTO 작성

kendrickhama committed:
♻️ [refactor] Role enum "ROLE_OOO" 로 수정

kendrickhama committed:
♻️ [refactor] replace로 변경

kendrickhama committed:
♻️ [refactor] 쿠키만료 시간 변경

kendrickhama committed:
♻️ [refactor] 컨트롤러와 로직분리를 위한 ReIssueService 작성

kendrickhama committed:
♻️ [refactor] @Value 메서드 밖으로 추출

kendrickhama committed:
♻️ [refactor] 이메일,닉네임,휴대번호 중복 체크, response에 isavaliable 담기위한 dto 생성, 및 controller, service 코드 수정

kendrickhama committed:
♻️ [refactor] ReissueService를 위한 errorcode 추가

kendrickhama committed:
Merge branch 'dev' into feat/spring-security-기능-구현-#51

kendrickhama committed:
merge 후 충돌난 부분 따로 추가

kendrickhama committed:
♻️ [refactor] 코드 리팩토링
1. UserService의 중복검증 코드 DTO 반환으로 인한
test 코드 리팩토링
2. UserRepositoryCustomImpl 의 queryDsl existByEmail 같은 메서드 존재하면 true로 변경

kendrickhama committed:
♻️ [refactor] CustomUserDetails에 Getter

kendrickhama committed:
Merge branch 'dev' into feat/spring-security-기능-구현-#51

kbyunghoon commented on src/main/java/com/investmetic/global/config/SecurityConfig.java:
```diff
@@ -1,37 +1,100 @@
+    @Bean
+    public static RoleHierarchy roleHierarchy() {
+        return RoleHierarchyImpl.fromHierarchy("""
+            SUPER_ADMIN > TRADER_ADMIN
+            SUPER_ADMIN > INVESTOR_ADMIN
+            TRADER_ADMIN > TRADER
+            INVESTOR_ADMIN > INVESTOR
```
이 부분 수정안된건가요?? 아니면 이대로 가는게 맞는건가요?

kbyunghoon commented on src/main/java/com/investmetic/global/security/filter/CustomLogoutFilter.java:
```diff
@@ -0,0 +1,109 @@
+            return;
+        }
+
+        //로그아웃 진행
+        //Refresh 토큰 DB에서 제거
+        redisUtil.deleteRefreshToken(username);
+
+        Cookie cookie = new Cookie("refresh", null);
+        cookie.setMaxAge(0);
+        cookie.setPath("/");
+
+        response.addCookie(cookie);
+        response.setStatus(HttpServletResponse.SC_OK);
```
로그아웃은 컨트롤러로 따로 빼는 게 좋지 않을까요?

kbyunghoon reviewed (changes requested):
리뷰 한번 확인해주세요

kendrickhama committed:
♻️ [refactor] Logout을 필터가 아닌 컨트롤러에서 처리하기 위한 컨트롤러와 서비스 추가

kbyunghoon commented on src/main/java/com/investmetic/global/security/filter/JWTFilter.java:
```diff
@@ -0,0 +1,86 @@
+
+        // 토큰 만료 여부 확인, 만료시 다음 필터로 넘기지 않음
+        try {
+            jwtUtil.isExpired(accessToken);
```
isExpired 만료 안 될 경우 `false`지만
만료 될 경우에는 `true`를 반환하는 건 없어보입니다.
이 경우 isExpired에서 예외처리에서 return true로 하면 isExpired를 호출하는 코드에선 try catch를 더 이상 사용할 필요가 없지 않을까요?

kendrickhama committed:
♻️ [refactor] ROLE_ENUM에 접두사 추가

kendrickhama committed:
♻️ [refactor] isExpired에 try-catch 처리 완료

kendrickhama committed:
♻️ [refactor] access,refresh key 이름에 접두사 추가

kendrickhama committed:
Merge branch 'dev' into feat/spring-security-기능-구현-#51

ziguin0925 commented on src/main/java/com/investmetic/domain/user/dto/response/TokenResponse.java:
```diff
@@ -0,0 +1,17 @@
+
+@Getter
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+public class TokenResponse {
```
해당 클래스도 사용하지 않는것 같은데 한번 확인 가능할까요?

ziguin0925 commented on src/main/java/com/investmetic/global/config/SecurityConfig.java:
```diff
@@ -1,37 +1,95 @@
         http
+                .cors(cors -> cors.configurationSource(request -> {
+                    CorsConfiguration config = new CorsConfiguration();
+                    config.setAllowedOrigins(List.of("http://localhost:3000"));
```
patch등 추가 부탁드릴게요

ziguin0925 reviewed (commented):
리뷰 남겼습니당

kendrickhama committed:
🐛 [fix] "ROLE_" 접두사 추가 후 접속 권한 오류 수정

kendrickhama committed:
🔥 [remove] 사용하지 않는 변수 및 메소드 삭제

kendrickhama committed:
♻️ [refactor] access token,refresh token 변수명 수정

kendrickhama committed:
♻️ [refactor] refresh token 오타 수정
